android/app/src/main/java/com/zulipmobile/TextCompressionModule.java

```diff
@@ -0,0 +1,81 @@
+package com.zulipmobile;
+
+import android.util.Base64;
+
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.Promise;
+
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+import android.util.Log;
+
+class TextCompressionModule extends ReactContextBaseJavaModule {
+  public TextCompressionModule(ReactApplicationContext reactContext) {
+    super(reactContext);
+  }
+
+  @Override
+  public String getName() {
+    return "TextCompressionModule";
+  }
+
+  // TODO: Experiment what value gives the best performance.
+  private int bufferSize = 8192;
+
+  @ReactMethod
+  public void compress(String input, Promise promise) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      Deflater deflater = new Deflater();
+      deflater.setInput(input.getBytes("UTF-8"));
+      deflater.finish();
+      byte[] buffer = new byte[bufferSize];
+      while (!deflater.finished()) {
+        int byteCount = deflater.deflate(buffer);
+        outputStream.write(buffer, 0, byteCount);
+      }
+      deflater.end();
+      outputStream.close();
+      // The RN bridge currently doesn't support sending binary data, so send the compressed
+      // output as a string. Also, out of unknown reasons, using ISO-8859-1 to directly encode
+      // the output as a string doesn't work. Encoding it as a Base64 string does work, but it
+      // is unclear why. Curiously, Base64.encodeToString itself uses ISO-8859-1.
```

Comment:
Style nit: for good readability, prose should be wrapped at a narrower width than the maximum line length allowed for code.

About 70 columns is a good limit for the width of prose. Comment lines generally start a bit indented, so you can go past column 70 of the line; in VS Code, I set
```
    "editor.rulers": [76, 100],
```
to wrap text at 76 total columns. This comment starts 9 columns in, so going as deep as 79 would be reasonable.

Same principle applies to the block comments in the JS file.